modules/xmpp/XmppConnection.js

```diff
@@ -377,18 +383,35 @@ export default class XmppConnection extends Listenable {
 
             logger.debug(`Scheduling next WebSocket keep-alive in ${intervalWithJitter}ms`);
 
-            this._wsKeepAlive = setTimeout(() => {
-                const url = this.service.replace('wss://', 'https://').replace('ws://', 'http://');
-
-                fetch(url).catch(
-                    error => {
-                        logger.error(`Websocket Keep alive failed for url: ${url}`, { error });
-                    })
-                    .then(() => this._maybeStartWSKeepAlive());
-            }, intervalWithJitter);
+            this._wsKeepAlive = setTimeout(this._checkShardAndMaybeScheduleKeepAlive, intervalWithJitter);
         }
     }
 
+    /**
+     * Do a http GET to the shard and if shard change will throw an event.
+     * Also schedules a future ws keep-alive.
+     * @private
+     */
+    _checkShardAndMaybeScheduleKeepAlive() {
+        const { shard } = this._options;
+        const url = this.service.replace('wss://', 'https://').replace('ws://', 'http://');
+
+        fetch(url)
+            .then(response => {
+                const responseShard = response.headers.get('x-jitsi-shard');
+
+                if (responseShard !== shard) {
+                    logger.error(
+                        `Detected that shard changed from ${shard} to ${responseShard}`);
+                    this.eventEmitter.emit(XmppConnection.Events.CONN_SHARD_CHANGED);
+                }
+            })
+            .catch(error => {
+                logger.error(`Websocket Keep alive failed for url: ${url}`, { error });
+            })
+            .then(() => this._maybeStartWSKeepAlive());
```

Comment:
I would remove `.then(() => this._maybeStartWSKeepAlive());` and make the method return a promise, this method should not care about the interval, maybe call it `keepAliveAndCheckShard`